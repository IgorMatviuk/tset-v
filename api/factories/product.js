const shortid = require('shortid')

function getImageMap(i) {
  const imgIndex = (i % 4) + 1

  const imgTmpl = {
    thumb_lg: `/img/previews/preview-${imgIndex}-large.jpg`,
    original_image_url: `/img/previews/preview-${imgIndex}-large.jpg`,
    thumb_md: `/img/previews/preview-${imgIndex}-medium.jpg`,
    thumb_sm: `/img/previews/preview-${imgIndex}-small.jpg`,
    thumb_xs: `/img/previews/preview-${imgIndex}-small.jpg`
  }

  return imgTmpl
}

function getPreviews() {
  return [
    {
      preview_type: 'image',
      title: 'some title',
      preview_xl: '/img/product/product__large-1.jpg',
      preview_lg: '/img/product/product__medium-1.jpg',
      original: '/img/product/product__medium-1.jpg'
    },
    {
      preview_type: 'image',
      title: 'some title',
      preview_xl: '/img/product/product__large-2.jpg',
      preview_lg: '/img/product/product__medium-2.jpg',
      original: '/img/product/product__medium-2.jpg'
    },
    {
      preview_type: 'image',
      title: 'some title',
      preview_xl: '/img/product/product__large-3.jpg',
      preview_lg: '/img/product/product__medium-3.jpg',
      original: '/img/product/product__medium-3.jpg'
    },
    {
      preview_type: 'image',
      title: 'some title',
      preview_xl: '/img/product/product__large-4.jpg',
      preview_lg: '/img/product/product__medium-4.jpg',
      original: '/img/product/product__medium-4.jpg'
    },
    {
      preview_type: 'image',
      title: 'some title',
      preview_xl: '/img/product/product__large-1.jpg',
      preview_lg: '/img/product/product__medium-1.jpg',
      original: '/img/product/product__medium-1.jpg'
    },
    {
      preview_type: 'video',
      video_url: '/video/product/video.mp4',
      poster_url: '/img/product/product__large-1.jpg'
    }
  ]
}

function getColorOptions(i) {
  const colors = [
    [
      { id: 111, label: 'Чёрный', slug: 'chyornyj', swatch_value: '#000000' },
      {
        id: 222,
        label: 'Оранжевый',
        slug: 'oranzhevy',
        swatch_value: '#E77D00'
      },
      {
        id: 333,
        label: 'Коричневый',
        slug: 'korichnevy',
        swatch_value: '#A26767'
      }
    ],
    [
      { id: 111, label: 'Чёрный', slug: 'chyornyj', swatch_value: '#000000' },
      { id: 222, label: 'Синий', slug: 'sinij', swatch_value: '#000F95' }
    ],
    [
      {
        id: 111,
        label: 'Бежевый',
        slug: 'bezhevyj',
        swatch_value: '#EEDBD5'
      },
      {
        id: 222,
        label: 'Бирюзовый',
        slug: 'biryuzovy',
        swatch_value: '#82BDCF'
      },
      { id: 333, label: 'Бардовый', slug: 'bardovy', swatch_value: '#9C1744' }
    ],
    [
      {
        id: 111,
        label: 'Бежевый',
        slug: 'bezhevyj',
        swatch_value: '#EEDBD5'
      },
      {
        id: 222,
        label: 'Бирюзовый',
        slug: 'biryuzovy',
        swatch_value: '#82BDCF'
      },
      { id: 333, label: 'Серый', slug: 'sery', swatch_value: '#C4C4C4' }
    ]
  ]
  return colors[i % colors.length]
}

function getSizeOptions() {
  return [
    { id: 444, label: 'XS', slug: 'xs', swatch_value: null },
    { id: 555, label: 'S', slug: 's', swatch_value: null },
    { id: 666, label: 'M', slug: 'm', swatch_value: null },
    { id: 777, label: 'L', slug: 'l', swatch_value: null }
  ]
}

function getProductName(i) {
  const names = [
    'Брюки кожаные с карманами',
    'Хлопковое платье',
    'Струящееся платье с цветочным принтом',
    'Трикотажное платье с поясом'
  ]

  return names[i % names.length]
}

function getStockStatus(i, context) {
  const isSale = context.discount_price(i) !== null
  const isNew = context.is_new(i)

  return isSale || isNew
}

function getDiscountPrice(i, type) {
  if (type === 'formatted') {
    return i % 2 === 0 ? '750 грн' : null
  } else {
    return i % 2 === 0 ? '750.0000' : null
  }
}

function getAttributeFamily() {
  return {
    description: {
      short_description: {
        value: 'Some description'
      }
    },
    spec: {
      composition: {
        code: 'composition',
        name: 'Ткань и состав',
        type: 'text',
        value: null
      },
      measurements: {
        code: 'measurements',
        name: 'Обмеры',
        type: 'text',
        value: null
      }
    },
    info: {
      care: {
        code: 'care',
        name: 'Уход за изделием',
        type: 'textarea',
        value: null
      }
    }
  }
}

function getVariantId() {
  return shortid.generate()
}

function getVariants(i, context) {
  return [
    {
      id: getVariantId(),
      type: 'simple',
      sku: 'dress-base-variant-1-7',
      name: getProductName(i),
      url_key: `strujashheesja-platje-s-cvetochnym-printom-base-variant-1-7-${i}`,
      price: '100.0000',
      formatted_price: '1500 грн',
      discount_price: getDiscountPrice(i, 'default'),
      formatted_discount_price: getDiscountPrice(i, 'formatted'),
      attributes: [
        {
          id: 7,
          code: 'color',
          locale: null,
          channel: null,
          value: 111,
          product_id: 2,
          attribute_id: 23,
          attribute: {
            id: 23,
            code: 'color',
            admin_name: 'Color',
            type: 'select',
            validation: '',
            position: 23,
            is_required: 1,
            is_unique: 0,
            value_per_locale: 0,
            value_per_channel: 0,
            is_filterable: 1,
            is_configurable: 1,
            is_user_defined: 1,
            is_visible_on_front: 1,
            created_at: '2020-02-25 19:52:47',
            updated_at: '2020-02-27 15:45:15',
            swatch_type: 'color',
            use_in_flat: 1,
            name: 'Color',
            translations: [
              {
                id: 23,
                locale: 'en',
                name: 'Color',
                attribute_id: 23
              }
            ]
          }
        },
        {
          id: 8,
          code: 'size',
          locale: null,
          channel: null,
          value: 444,
          product_id: 2,
          attribute_id: 24,
          attribute: {
            id: 24,
            code: 'size',
            admin_name: 'Size',
            type: 'select',
            validation: null,
            position: 24,
            is_required: 0,
            is_unique: 0,
            value_per_locale: 0,
            value_per_channel: 0,
            is_filterable: 1,
            is_configurable: 1,
            is_user_defined: 1,
            is_visible_on_front: 0,
            created_at: '2020-02-25 19:52:47',
            updated_at: '2020-02-25 19:52:47',
            swatch_type: null,
            use_in_flat: 1,
            name: 'Size',
            translations: [
              {
                id: 24,
                locale: 'en',
                name: 'Size',
                attribute_id: 24
              }
            ]
          }
        }
      ],
      attribute_family: getAttributeFamily(),
      images: [],
      base_image: getImageMap(i),
      previews: getPreviews(),
      is_new: null,
      is_featured: null,
      in_stock: getStockStatus(i, context),
      variants: [],
      short_description: null,
      description: null,
      is_saved: false,
      created_at: '2020-02-27 15:23:05',
      updated_at: '2020-02-27 15:23:05'
    },
    {
      id: getVariantId(),
      type: 'simple',
      sku: 'dress-base-variant-2-7',
      name: getProductName(i + 1),
      url_key: `strujashheesja-platje-s-cvetochnym-printom-base-variant-2-7-${i}`,
      price: '100.0000',
      formatted_price: '1500 грн',
      discount_price: getDiscountPrice(i, 'default'),
      formatted_discount_price: getDiscountPrice(i, 'formatted'),
      attributes: [
        {
          id: 14,
          code: 'color',
          locale: null,
          channel: null,
          value: 222,
          product_id: 3,
          attribute_id: 23,
          attribute: {
            id: 23,
            code: 'color',
            admin_name: 'Color',
            type: 'select',
            validation: '',
            position: 23,
            is_required: 1,
            is_unique: 0,
            value_per_locale: 0,
            value_per_channel: 0,
            is_filterable: 1,
            is_configurable: 1,
            is_user_defined: 1,
            is_visible_on_front: 1,
            created_at: '2020-02-25 19:52:47',
            updated_at: '2020-02-27 15:45:15',
            swatch_type: 'color',
            use_in_flat: 1,
            name: 'Color',
            translations: [
              {
                id: 23,
                locale: 'en',
                name: 'Color',
                attribute_id: 23
              }
            ]
          }
        },
        {
          id: 15,
          code: 'size',
          locale: null,
          channel: null,
          value: 444,
          product_id: 3,
          attribute_id: 24,
          attribute: {
            id: 24,
            code: 'size',
            admin_name: 'Size',
            type: 'select',
            validation: null,
            position: 24,
            is_required: 0,
            is_unique: 0,
            value_per_locale: 0,
            value_per_channel: 0,
            is_filterable: 1,
            is_configurable: 1,
            is_user_defined: 1,
            is_visible_on_front: 0,
            created_at: '2020-02-25 19:52:47',
            updated_at: '2020-02-25 19:52:47',
            swatch_type: null,
            use_in_flat: 1,
            name: 'Size',
            translations: [
              {
                id: 24,
                locale: 'en',
                name: 'Size',
                attribute_id: 24
              }
            ]
          }
        }
      ],
      attribute_family: getAttributeFamily(),
      images: [],
      base_image: getImageMap(i + 1),
      previews: getPreviews(),
      is_new: null,
      is_featured: null,
      in_stock: getStockStatus(i, context),
      variants: [],
      short_description: null,
      description: null,
      is_saved: false,
      created_at: '2020-02-27 15:23:05',
      updated_at: '2020-02-27 15:23:05'
    },
    {
      id: getVariantId(),
      type: 'simple',
      sku: 'dress-base-variant-1-8',
      name: getProductName(i),
      url_key: `strujashheesja-platje-s-cvetochnym-printom-base-variant-1-8-${i}`,
      price: '100.0000',
      formatted_price: '1500 грн',
      discount_price: getDiscountPrice(i, 'default'),
      formatted_discount_price: getDiscountPrice(i, 'formatted'),
      attributes: [
        {
          id: 21,
          code: 'color',
          locale: null,
          channel: null,
          value: 111,
          product_id: 4,
          attribute_id: 23,
          attribute: {
            id: 23,
            code: 'color',
            admin_name: 'Color',
            type: 'select',
            validation: '',
            position: 23,
            is_required: 1,
            is_unique: 0,
            value_per_locale: 0,
            value_per_channel: 0,
            is_filterable: 1,
            is_configurable: 1,
            is_user_defined: 1,
            is_visible_on_front: 1,
            created_at: '2020-02-25 19:52:47',
            updated_at: '2020-02-27 15:45:15',
            swatch_type: 'color',
            use_in_flat: 1,
            name: 'Color',
            translations: [
              {
                id: 23,
                locale: 'en',
                name: 'Color',
                attribute_id: 23
              }
            ]
          }
        },
        {
          id: 22,
          code: 'size',
          locale: null,
          channel: null,
          value: 555,
          product_id: 4,
          attribute_id: 24,
          attribute: {
            id: 24,
            code: 'size',
            admin_name: 'Size',
            type: 'select',
            validation: null,
            position: 24,
            is_required: 0,
            is_unique: 0,
            value_per_locale: 0,
            value_per_channel: 0,
            is_filterable: 1,
            is_configurable: 1,
            is_user_defined: 1,
            is_visible_on_front: 0,
            created_at: '2020-02-25 19:52:47',
            updated_at: '2020-02-25 19:52:47',
            swatch_type: null,
            use_in_flat: 1,
            name: 'Size',
            translations: [
              {
                id: 24,
                locale: 'en',
                name: 'Size',
                attribute_id: 24
              }
            ]
          }
        }
      ],
      attribute_family: getAttributeFamily(),
      images: [],
      base_image: getImageMap(i),
      previews: getPreviews(),
      is_new: null,
      is_featured: null,
      in_stock: getStockStatus(i, context),
      variants: [],
      short_description: null,
      description: null,
      is_saved: false,
      created_at: '2020-02-27 15:23:06',
      updated_at: '2020-02-27 15:23:06'
    },
    {
      id: getVariantId(),
      type: 'simple',
      sku: 'dress-base-variant-2-8',
      name: getProductName(i + 1),
      url_key: `strujashheesja-platje-s-cvetochnym-printom-base-variant-2-8-${i}`,
      price: '100.0000',
      formatted_price: '1500 грн',
      discount_price: getDiscountPrice(i, 'default'),
      formatted_discount_price: getDiscountPrice(i, 'formatted'),
      attributes: [
        {
          id: 28,
          code: 'color',
          locale: null,
          channel: null,
          value: 222,
          product_id: 5,
          attribute_id: 23,
          attribute: {
            id: 23,
            code: 'color',
            admin_name: 'Color',
            type: 'select',
            validation: '',
            position: 23,
            is_required: 1,
            is_unique: 0,
            value_per_locale: 0,
            value_per_channel: 0,
            is_filterable: 1,
            is_configurable: 1,
            is_user_defined: 1,
            is_visible_on_front: 1,
            created_at: '2020-02-25 19:52:47',
            updated_at: '2020-02-27 15:45:15',
            swatch_type: 'color',
            use_in_flat: 1,
            name: 'Color',
            translations: [
              {
                id: 23,
                locale: 'en',
                name: 'Color',
                attribute_id: 23
              }
            ]
          }
        },
        {
          id: 29,
          code: 'size',
          locale: null,
          channel: null,
          value: 555,
          product_id: 5,
          attribute_id: 24,
          attribute: {
            id: 24,
            code: 'size',
            admin_name: 'Size',
            type: 'select',
            validation: null,
            position: 24,
            is_required: 0,
            is_unique: 0,
            value_per_locale: 0,
            value_per_channel: 0,
            is_filterable: 1,
            is_configurable: 1,
            is_user_defined: 1,
            is_visible_on_front: 0,
            created_at: '2020-02-25 19:52:47',
            updated_at: '2020-02-25 19:52:47',
            swatch_type: null,
            use_in_flat: 1,
            name: 'Size',
            translations: [
              {
                id: 24,
                locale: 'en',
                name: 'Size',
                attribute_id: 24
              }
            ]
          }
        }
      ],
      images: [],
      base_image: getImageMap(i + 1),
      previews: getPreviews(),
      is_new: null,
      is_featured: null,
      in_stock: getStockStatus(i, context),
      variants: [],
      short_description: null,
      description: null,
      is_saved: false,
      created_at: '2020-02-27 15:23:06',
      updated_at: '2020-02-27 15:23:06'
    },
    {
      id: getVariantId(),
      type: 'simple',
      sku: 'dress-base-variant-2-9',
      name: getProductName(i + 1),
      url_key: `strujashheesja-platje-s-cvetochnym-printom-base-variant-2-9-${i}`,
      price: '100.0000',
      formatted_price: '1500 грн',
      discount_price: getDiscountPrice(i, 'default'),
      formatted_discount_price: getDiscountPrice(i, 'formatted'),
      attributes: [
        {
          id: 28,
          code: 'color',
          locale: null,
          channel: null,
          value: 222,
          product_id: 5,
          attribute_id: 23,
          attribute: {
            id: 23,
            code: 'color',
            admin_name: 'Color',
            type: 'select',
            validation: '',
            position: 23,
            is_required: 1,
            is_unique: 0,
            value_per_locale: 0,
            value_per_channel: 0,
            is_filterable: 1,
            is_configurable: 1,
            is_user_defined: 1,
            is_visible_on_front: 1,
            created_at: '2020-02-25 19:52:47',
            updated_at: '2020-02-27 15:45:15',
            swatch_type: 'color',
            use_in_flat: 1,
            name: 'Color',
            translations: [
              {
                id: 23,
                locale: 'en',
                name: 'Color',
                attribute_id: 23
              }
            ]
          }
        },
        {
          id: 29,
          code: 'size',
          locale: null,
          channel: null,
          value: 666,
          product_id: 6,
          attribute_id: 24,
          attribute: {
            id: 24,
            code: 'size',
            admin_name: 'Size',
            type: 'select',
            validation: null,
            position: 24,
            is_required: 0,
            is_unique: 0,
            value_per_locale: 0,
            value_per_channel: 0,
            is_filterable: 1,
            is_configurable: 1,
            is_user_defined: 1,
            is_visible_on_front: 0,
            created_at: '2020-02-25 19:52:47',
            updated_at: '2020-02-25 19:52:47',
            swatch_type: null,
            use_in_flat: 1,
            name: 'Size',
            translations: [
              {
                id: 24,
                locale: 'en',
                name: 'Size',
                attribute_id: 24
              }
            ]
          }
        }
      ],
      images: [],
      base_image: getImageMap(i + 1),
      previews: getPreviews(),
      is_new: null,
      is_featured: null,
      in_stock: getStockStatus(i, context),
      variants: [],
      short_description: null,
      description: null,
      is_saved: false,
      created_at: '2020-02-27 15:23:06',
      updated_at: '2020-02-27 15:23:06'
    }
  ]
}

const productAttrs = {
  id(i) {
    return i + 1
  },

  name(i) {
    return getProductName(i)
  },

  url_key(i) {
    const keys = [
      'brjuki-kozhanye-s-karmanami',
      'hlopkovoe-platje',
      'strujashheesja-platje-s-cvetochnym-printom',
      'trikotazhnoe-platje-s-pojasom'
    ]

    return `${keys[i % keys.length]}-${i + 1}`
  },

  sku(i) {
    return `VR0${i + 1}-0100668`
  },

  price(i) {
    const prices = ['1150.0000', '950.0000', '750.0000', '550.0000']

    return prices[i % prices.length]
  },

  formatted_price(i) {
    const formatedPrices = ['1150 грн', '950 грн', '750 грн', '550 грн']

    return formatedPrices[i % formatedPrices.length]
  },

  discount_price(i) {
    const discountPrice = '750.0000'

    return i % 2 === 0 ? discountPrice : null
  },

  formatted_discount_price(i) {
    const formatedFiscountPrice = '750 грн'

    return i % 2 === 0 ? formatedFiscountPrice : null
  },

  attributes(i) {
    const attrList = [
      [
        {
          id: 11,
          code: 'color',
          name: 'Цвет',
          options: getColorOptions(i)
        },
        {
          id: 22,
          code: 'size',
          name: 'Размер',
          options: getSizeOptions()
        }
      ],
      [
        {
          id: 11,
          code: 'color',
          name: 'Цвет',
          options: getColorOptions(i)
        },
        {
          id: 22,
          code: 'size',
          name: 'Размер',
          options: getSizeOptions()
        }
      ],
      [
        {
          id: 11,
          code: 'color',
          name: 'Цвет',
          options: getColorOptions(i)
        },
        {
          id: 22,
          code: 'size',
          name: 'Размер',
          options: getSizeOptions()
        }
      ],
      [
        {
          id: 11,
          code: 'color',
          name: 'Цвет',
          options: getColorOptions(i)
        },
        {
          id: 22,
          code: 'size',
          name: 'Размер',
          options: getSizeOptions()
        }
      ]
    ]

    return attrList[i % attrList.length]
  },

  attribute_family() {
    return getAttributeFamily()
  },

  images(i) {
    const imgMap = getImageMap(i)

    return [imgMap]
  },

  base_image(i) {
    return getImageMap(i)
  },

  previews() {
    return getPreviews()
  },

  variants(i) {
    return getVariants(i, this)
  },

  is_new(i) {
    const isSale = this.discount_price(i) !== null
    return !isSale && (i + 1) % 3 === 0
  },

  in_stock(i) {
    return getStockStatus(i, this)
  },

  reviews() {
    return []
  }
}

module.exports = productAttrs
