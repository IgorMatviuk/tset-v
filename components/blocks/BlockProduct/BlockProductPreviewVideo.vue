<template>
  <div class="preview-video">
    <client-only>
      <LazyVideo
        :src="video.video_url"
        :poster="video[posterUrlKey]"
        :attrs="attrs"
      />
    </client-only>
  </div>
</template>

<script>
export default {
  props: {
    video: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      attrs: null,
      mqlLg: null
    }
  },
  computed: {
    posterUrlKey() {
      return this.mqlLg && this.mqlLg.matches ? 'preview_lg' : 'preview_xl'
    }
  },
  mounted() {
    this.mqlLg = matchMedia('(max-width: 1439px)')

    this.attrs = {
      preload: 'none',
      muted: true,
      loop: true,
      autoplay: this.getAutoplay(),
      playsinline: true,
      controls: true,
      onplaying: 'this.controls=false',
      class: 'preview-video__item',
      'data-object-fit': 'cover'
    }
  },
  methods: {
    getAutoplay() {
      const mqlLg = matchMedia('(min-width: 768px)')
      return mqlLg.matches
    }
  }
}
</script>

<style lang="scss">
.preview-video {
  position: absolute;
  width: 100%;
  height: 0;
  padding-bottom: $preview-aspect-ratio;

  &__item {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    object-fit: cover;
    width: 100%;
    min-width: 100%;
    height: 100%;
  }

  /* @include media-breakpoint-down(xl) {
    padding-bottom: 126.74%;
  }

  @include media-breakpoint-down(lg) {
    padding-bottom: 126.54%;
  }

  @include media-breakpoint-down(md) {
    padding-bottom: 126.69%;
  }

  @include media-breakpoint-down(sm) {
    min-height: 76.5vh;
    padding-bottom: 135%;
  }

  @media only screen and (max-width: 767.98px) and (orientation: landscape) {
    padding-bottom: 126.68%;
  } */
}
</style>
