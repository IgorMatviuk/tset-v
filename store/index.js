import { actionTypes as catalogActionTypes } from '@/store/catalog'

export const mutationTypes = {
  SET_LANG: 'SET_LANG',
  SET_MENUS: 'SET_MENUS',
  SET_PAGE_TYPE: 'SET_PAGE_TYPE',
  SET_BREADCRUMB: 'SET_BREADCRUMB',
  SET_CONTACTS: 'SET_CONTACTS',
  SET_GLOBALS: 'SET_GLOBALS'
}

export const actionTypes = {
  GET_MENUS: 'GET_MENUS',
  GET_CONTACTS: 'GET_CONTACTS',
  GET_BANNER: 'GET_BANNER',
  GET_CLIENT_LOCATION: 'GET_CLIENT_LOCATION'
}

export const state = () => {
  return {
    locale: 'ru',
    locales: ['uk', 'ru'],
    fallbackLocale: 'ru',
    pageType: null,
    breadcrumb: [],
    menus: [],
    globals: {
      countryCode: 'UA',
      banner: null
    },
    contacts: {
      phone: '0 800 332 512',
      email: 'office@virna.com.ua',
      schedule: 'с 9:00 - 18:00',
      socialLinks: [
        { name: 'instagram', url: 'https://www.instagram.com/' },
        { name: 'facebook', url: 'https://facebook.com/' },
        { name: 'youtube', url: 'https://www.youtube.com/' }
      ]
    }
  }
}

export const mutations = {
  [mutationTypes.SET_LANG](state, locale) {
    state.locale = locale
  },
  [mutationTypes.SET_MENUS](state, menus) {
    state.menus = menus
  },
  [mutationTypes.SET_PAGE_TYPE](state, pageType) {
    state.pageType = pageType
  },
  [mutationTypes.SET_BREADCRUMB](state, breadcrumb) {
    state.breadcrumb = breadcrumb
  },
  [mutationTypes.SET_CONTACTS](state, contacts) {
    state.contacts = contacts
  },
  [mutationTypes.SET_GLOBALS](state, { key, value }) {
    state.globals[key] = value
  }
}

export const actions = {
  async [actionTypes.GET_MENUS]({ commit }) {
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: menus } = await this.$axios.$get(`${STAGE_API_URL}/api/menus`)
    commit(mutationTypes.SET_MENUS, menus)
  },
  async [actionTypes.GET_CONTACTS]({ commit }) {
    const STAGE_API_URL = 'https://back.virna.com.ua'
    const { data: contacts } = await this.$axios.$get(
      `${STAGE_API_URL}/api/contacts`
    )
    commit(mutationTypes.SET_CONTACTS, contacts)
  },
  async [actionTypes.GET_BANNER]({ commit }) {
    const { data: banner } = await this.$axios.$get('/api/banner')
    const defaultBanner = {
      key: 'banner',
      value: banner
    }

    commit(mutationTypes.SET_GLOBALS, defaultBanner)
  },
  async [actionTypes.GET_CLIENT_LOCATION]({ commit }) {
    const locationString = await this.$axios.$get('https://ip2c.org/s')
    const [, countryCode] = locationString.split(';')
    const defaultCountry = {
      key: 'countryCode',
      value: countryCode
    }

    commit(mutationTypes.SET_GLOBALS, defaultCountry)
  },
  async nuxtServerInit({ dispatch }) {
    await Promise.all([
      dispatch(actionTypes.GET_MENUS),
      dispatch(actionTypes.GET_CONTACTS),
      dispatch(`catalog/${catalogActionTypes.GET_CATEGORIES}`)
    ])
  }
}

/**
 * Get menum by type
 * @param {object} state
 * @param {('header'|'footer'|'catalog'|'promo')} type
 */
function getMenuByType(state, type) {
  const { menus } = state
  return menus.find((menu) => menu.menu_type === type)
}

export const getters = {
  mobileMenu(state) {
    return getMenuByType(state, 'mobile')
  },
  desktopMenu(state) {
    return getMenuByType(state, 'header')
  },
  promoMenu(state) {
    return getMenuByType(state, 'promo')
  },
  topMenu(state) {
    return getMenuByType(state, 'top')
  },
  catalogMenu(state) {
    return getMenuByType(state, 'catalog')
  },
  footerMenu(state) {
    return getMenuByType(state, 'footer')
  }
}
