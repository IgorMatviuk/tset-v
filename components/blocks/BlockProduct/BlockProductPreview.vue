<template>
  <div
    class="product-preview-card"
    @mouseover="activateHoverZoom"
    @click="showImageViewer"
  >
    <template v-if="isImagePreview">
      <ImageLoader ref="imageWrapper" class="product-preview-card__img-wrapper">
        <picture>
          <source
            :data-srcset="item.preview_lg | webp"
            srcset="/img/stub.png"
            media="(max-width: 1439px)"
            type="image/webp"
          />
          <source
            :data-srcset="item.preview_lg"
            srcset="/img/stub.png"
            media="(max-width: 1439px)"
            type="image/jpeg"
          />
          <source
            :data-srcset="item.preview_xl | webp"
            srcset="/img/stub.png"
            type="image/webp"
          />
          <img
            :data-src="item.preview_xl"
            :title="item.title"
            :alt="item.alt"
            src="/img/stub.png"
            class="product-preview-card__img lazyload"
            loading="lazy"
            data-object-fit="cover"
          />
        </picture>
        <ZoomedImage v-if="hoverZoomIsActive" :image-url="item.preview_xl" />
      </ImageLoader>
    </template>
    <template v-else>
      <BlockProductPreviewVideo :video="item" />
    </template>
  </div>
</template>

<script>
import PinchZoom from 'pinch-zoom-js'

import BlockProductPreviewVideo from './BlockProductPreviewVideo'

import ImageLoader from '@/components/shared/ImageLoader'
import ZoomedImage from '@/components/shared/ZoomedImage'

import { mutationTypes as productMutationsTypes } from '@/store/product'

import imageNextGen from '@/mixins/image-next-gen'

export default {
  components: {
    ImageLoader,
    ZoomedImage,
    BlockProductPreviewVideo
  },
  mixins: [imageNextGen],
  props: {
    item: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      hoverZoomIsActive: false,
      pinchZoom: null
    }
  },
  computed: {
    isImagePreview() {
      return this.item.preview_type === 'image'
    }
  },
  mounted() {
    if (this.$refs.imageWrapper) {
      this.setupPinchZoom()
    }

    window.addEventListener('resize', this.destroyHoverZoom)
  },
  beforeDestroy() {
    if (this.pinchZoom) {
      this.pinchZoom.disable()
      this.pinchZoom = null
    }

    window.removeEventListener('resize', this.destroyHoverZoom)
  },
  methods: {
    setupPinchZoom() {
      const element = this.$refs.imageWrapper.$el
      const mqlLg = window.matchMedia('(max-width: 1199.98px)')

      /* istanbul ignore else */
      if (element && mqlLg.matches) {
        this.pinchZoom = new PinchZoom(element, { draggableUnzoomed: false })
      }
    },
    destroyHoverZoom() {
      this.hoverZoomIsActive = false
    },
    activateHoverZoom() {
      const mqlXl = window.matchMedia('(min-width: 1200px)')

      /* istanbul ignore else */
      if (mqlXl.matches) {
        this.hoverZoomIsActive = true
      }
    },
    showImageViewer() {
      /* istanbul ignore else */
      if (this.isImagePreview) {
        this.$store.commit(
          `product/${productMutationsTypes.SET_ACTIVE_PREVIEW}`,
          this.item
        )
        this.$store.commit(
          `product/${productMutationsTypes.SET_IMAGE_VIEWER_VISIBILITY_TOGGLE}`,
          true
        )
      }
    }
  }
}
</script>

<style lang="scss">
.product-preview-card {
  position: relative;
  height: 1px;
  padding-bottom: $preview-aspect-ratio;
  overflow: hidden;
  transform: translateZ(0);
  backface-visibility: hidden;

  &__img {
    width: 100%;
    height: 100%;
    object-fit: cover;
  }

  .pinch-zoom-container {
    position: absolute !important;
    width: 100%;
  }

  .zoomed {
    opacity: 0;
    transition: opacity 0.5s;
  }

  &:hover .zoomed {
    opacity: 1;
  }

  @include media-breakpoint-down(sm) {
    position: relative;
    width: 100%;
    height: 0;

    &__img-wrapper {
      width: 100%;
      height: 100%;
    }
  }
}
</style>
