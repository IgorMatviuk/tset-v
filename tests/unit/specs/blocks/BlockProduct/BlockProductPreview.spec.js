import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'

import BlockProductPreview from '@/components/blocks/BlockProduct/BlockProductPreview'
import BlockProductPreviewVideo from '@/components/blocks/BlockProduct/BlockProductPreviewVideo'

import { mutations as productMutations } from '@/store/product'

import createFactory from '@/api/helpers/factory'
import productAttrs from '@/api/factories/product'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('BlockProductPreview', () => {
  let imagePreview
  let videoPreview
  let store

  function createMatchMediaMock(matches) {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches,
        media: query,
        addListener: jest.fn().mockImplementation((callback) => callback()),
        removeListener: jest.fn().mockImplementation((callback) => callback()),
        addEventListener: jest.fn().mockImplementation((callback) => callback())
      }))
    })
  }

  beforeEach(() => {
    const products = createFactory(productAttrs, 12)
    const regularProduct = products.find((product) => product.in_stock)

    imagePreview = regularProduct.previews.find(
      (preview) => preview.preview_type === 'image'
    )
    videoPreview = regularProduct.previews.find(
      (preview) => preview.preview_type === 'video'
    )

    store = new Vuex.Store({
      modules: {
        product: {
          namespaced: true,
          state: {
            activeVariant: {},
            imageViewerVisibility: false,
            activePreview: {}
          },
          mutations: { ...productMutations }
        }
      }
    })
  })

  test('renders component itself', () => {
    createMatchMediaMock(true)

    const wrapper = shallowMount(BlockProductPreview, {
      propsData: {
        item: {}
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders image preview', () => {
    createMatchMediaMock(true)

    const wrapper = shallowMount(BlockProductPreview, {
      propsData: {
        item: imagePreview
      }
    })

    expect(wrapper.vm.pinchZoom).not.toBeNull()

    wrapper.destroy()

    expect(wrapper.vm.pinchZoom).toBeNull()
  })

  test('renders video preview', () => {
    createMatchMediaMock(true)

    const wrapper = shallowMount(BlockProductPreview, {
      propsData: {
        item: videoPreview
      }
    })

    expect(wrapper.findComponent(BlockProductPreviewVideo).exists()).toBe(true)
    expect(wrapper.vm.pinchZoom).toBeNull()

    wrapper.destroy()

    expect(wrapper.vm.pinchZoom).toBeNull()
  })

  test('initialize hover zoom', () => {
    createMatchMediaMock(true)

    const wrapper = shallowMount(BlockProductPreview, {
      localVue,
      store,
      propsData: {
        item: imagePreview
      }
    })

    const mouseoverEvent = new MouseEvent('mouseover')
    wrapper.element.dispatchEvent(mouseoverEvent)

    expect(wrapper.vm.hoverZoomIsActive).toBe(true)
  })

  test('update hover zoom on resize window', () => {
    createMatchMediaMock(true)

    const wrapper = shallowMount(BlockProductPreview, {
      localVue,
      store,
      propsData: {
        item: imagePreview
      }
    })

    const resizeEvent = new Event('resize')
    window.dispatchEvent(resizeEvent)

    expect(wrapper.vm.hoverZoomIsActive).toBe(false)
  })

  test('show image preview', () => {
    createMatchMediaMock(true)

    const wrapper = shallowMount(BlockProductPreview, {
      localVue,
      store,
      propsData: {
        item: imagePreview
      }
    })

    wrapper.trigger('click')

    expect(store.state.product.activePreview).toEqual(imagePreview)
    expect(store.state.product.imageViewerVisibility).toBe(true)
  })
})
