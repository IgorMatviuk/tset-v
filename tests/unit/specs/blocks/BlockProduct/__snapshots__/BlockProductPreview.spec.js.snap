// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockProductPreview renders component itself 1`] = `
<div
  class="product-preview-card"
>
  <blockproductpreviewvideo-stub
    video="[object Object]"
  />
</div>
`;
